rust/kernel/debugfs.rs

```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Rust implementation of `debugfs`.
+//!
+//! This module allows Rust kernel modules to create directories and files in
+//! `/debugfs`.
+//!
+//! C header: [`include/linux/debugfs.h`](../../../include/linux/debugfs.h)
+//!
+//! Reference: <https://www.kernel.org/doc/html/latest/filesystems/debugfs.html>
+
+use alloc::{boxed::Box, vec::Vec};
+use core::{
+    marker::{PhantomData, Sync},
+    ptr,
+};
+
+use crate::{bindings, c_types, error, str::CStr, types::PointerWrapper, Result};
+
+/// An `inode` for a directory in debugfs.
+pub struct DebugFsDirectory {
+    dentry: *mut bindings::dentry,
+    has_parent: bool,
+    file_children: Vec<Box<dyn Sync>>,
+    dir_children: Vec<DebugFsDirectory>,
```

Comment:
Creating a hierarchy that is recursively dropped in dangerous as it may lead to overflowing the call stack. Is there anything [that I'm failing to see] that prevents recursive calls to `drop` overflowing the stack? It seems like we need a strategy to clean these up without using recursion to traverse all the elements.